### PR TITLE
Screensaver on Kindle without Special Offers

### PR DESCRIPTION
**Pull request by robert00s**
Similar to #2734 but screensaver is available only on devices without Special Ofers.
@NiLuJe Thanks for advice.
Tested on my Kindle PW2 without SO (screensaver works) and PW3 with SO (screensaver isn't available).

### CODE REVIEW

File: frontend/ui/screensaver.lua [hunk 1]
```diff
@@ -147,4 +147,25 @@ function Screensaver:close()
...
+    if not lipc then
+        logger.warn("could not load liblibclua")
+        return false
+    end
+    local lipc_handle = lipc.init("com.github.koreader.device")
+    if not lipc_handle then
+        logger.warn("could not get lipc handle")
+        return false
+    end
+    local so = false
```
Comment (Frenzie) [inline frontend/ui/screensaver.lua]: Shouldn't the logic be the reverse from a better safe than sorry perspective?

File: frontend/ui/screensaver.lua [hunk 2]
```diff
@@ -147,4 +147,25 @@ function Screensaver:close()
     end
 end
 
+function Screensaver:isKindleSpecialOffers()
+    -- Look at the current blanket modules to see if the SO screensavers are enabled...
+    local lipc = require("liblipclua")
+    if not lipc then
+        logger.warn("could not load liblibclua")
+        return false
```
Comment (Frenzie) [inline frontend/ui/screensaver.lua]: Same for these.

File: frontend/ui/screensaver.lua [hunk 3]
```diff
@@ -147,4 +147,25 @@ function Screensaver:close()
...
+    end
+    local lipc_handle = lipc.init("com.github.koreader.device")
+    if not lipc_handle then
+        logger.warn("could not get lipc handle")
+        return false
```
Comment (Frenzie) [inline frontend/ui/screensaver.lua]: And this one.

Comment (robert00s) [inline frontend/ui/screensaver.lua]: Do you mean change false to true and change function name to eg isKindleWithoutSpecialOffers?

Comment (Frenzie) [inline frontend/ui/screensaver.lua]: I know very little about Kindles and even less about those special offers, but from the discussion I got the impression that a whitelist would be better than a blacklist. I think the logic at intoScreensaver() is fine but if there is no better way to do it I literally suggest flipping all the true/false values.

```lua
    if not string.find(loaded_blanket_modules, "ad_screensaver") then
        so = false
    end
```

That still seems rather broad, but at least it wouldn't be a cavalier "oh hey, we had a tiny issue with this lipc thingy so let's display the screensaver."

But I burned my hands on merging the last PR. Let's just see what @NiLuJe thinks. :-)

**Comment by NiLuJe** [comment]
Yeah, agreed, assuming that it's SO unless proven otherwise seems like a safer bet ;).

File: frontend/device/kindle/device.lua [hunk 4]
```diff
@@ -46,6 +46,9 @@ function Kindle:usbPlugIn()
 end
 
 function Kindle:intoScreenSaver()
+    if not require("ui/screensaver"):isKindleSpecialOffers() and require("ui/screensaver").isUsingBookCover then
```
Comment (houqp) [inline frontend/device/kindle/device.lua]: if you are doing multiple require here, might as well require once as save it to a local var for reuse:

```lua
local Screensaver = require("ui/screensaver")
if not self:hasSpecialOffers() and not Screensaver:isUsingBookCover() then
    Screensaver:show("suspend")
end
```

One tip I would like to point out is the coding style convention: we use `under_score_name` for variables and `camelCaseName` for functions. With that in mind, it's easy to spot that `require("ui/screensaver").isUsingBookCover` is a typo ;)

File: frontend/apps/filemanager/filemanagermenu.lua [hunk 5]
```diff
@@ -96,7 +96,7 @@ function FileManagerMenu:setUpdateItemTable()
             G_reader_settings:flush()
         end
     }
-    if Device.isKobo() then
+    if Device.isKobo() or (Device.isKindle() and not Screensaver:isKindleSpecialOffers()) then
```
Comment (houqp) [inline frontend/apps/filemanager/filemanagermenu.lua]: For better readability and efficiency, I suggest we create a device method specifically for this purpose. So in `frontend/device/generic/device.lua`, we can define a method called something like `supportsScreensaver()` which returns `false` by default. Then override this method in `frontend/device/kobo/device.lua` and `frontend/device/kindle/device.lua`.

File: frontend/ui/screensaver.lua (hunk 3 above)
Comment (houqp) [inline frontend/ui/screensaver.lua]: I agree, we should return true in all the error handling code path, i.e. only show screen saver when we are absolutely sure it's safe to do so. For example, a new version of kindle firmware upgrade might cause the lipc call to fail on devices with special offers, in that case, we want to make sure the screen saver is still not displayed.

File: frontend/ui/screensaver.lua (hunk 1 above)
Comment (houqp) [inline frontend/ui/screensaver.lua]: nitpick, for boolean variables, the coding convention is to prefix it with `is_`.

File: frontend/ui/screensaver.lua [hunk 6]
```diff
@@ -147,4 +147,25 @@ function Screensaver:close()
     end
 end
 
+function Screensaver:isKindleSpecialOffers()
```
Comment (houqp) [inline frontend/ui/screensaver.lua]: This is a device specific function, so it should live in the device module instead. Screensaver is a UI component that should be decoupled from device model logic.

**Comment by houqp** [comment]
Thanks @robert00s for the new patch!

File: frontend/apps/reader/modules/readermenu.lua [hunk 7]
```diff
@@ -145,6 +145,26 @@ function ReaderMenu:setUpdateItemTable()
                 }
             }
         }
+        elseif Device:isKindle() and Device:supportsScreensaver() then
+        self.menu_items.screensaver = {
+            text = _("Screensaver"),
+            sub_item_table = {
+                {
+                    text = _("Use book's cover as screensaver"),
```
Comment (houqp) [inline frontend/apps/reader/modules/readermenu.lua]: Can we keep this text consistent with what's in filemanager menu? It should be `_("Use last book's cover as screensaver")`.

File: frontend/apps/reader/modules/readermenu.lua [hunk 8]
```diff
@@ -145,6 +145,26 @@ function ReaderMenu:setUpdateItemTable()
                 }
             }
         }
+        elseif Device:isKindle() and Device:supportsScreensaver() then
```
Comment (houqp) [inline frontend/apps/reader/modules/readermenu.lua]: I think it's fine to enable this option for other devices too.

File: frontend/device/kindle/device.lua [hunk 9]
```diff
@@ -34,6 +34,40 @@ function Kindle:initNetworkManager(NetworkMgr)
     end
 end
 
+--[[
+Test if a kindle device has Special Offers
+--]]
+local function isSpecialOffers()
+    -- Look at the current blanket modules to see if the SO screensavers are enabled...
+    local lipc = require("liblipclua")
```
Comment (Frenzie) [inline frontend/device/kindle/device.lua]: Might need a pcall or something, see #3018

**Pull request merged**